src/tox/package.py

```diff
@@ -27,6 +28,22 @@ def get_package(session):
     if config.skipsdist:
         report.info("skipping sdist step")
         return None
+    lock_file = str(
+        session.config.toxworkdir.join("{}.lock".format(session.config.isolated_build_env))
+    )
+    lock = FileLock(lock_file)
+    try:
+        try:
+            lock.acquire(0.0001)
+        except Timeout:
+            report.verbosity0("lock file {} present, will block until released".format(lock_file))
+            lock.acquire()
+        return acquire_package(config, report, session)
```

Comment:
The problem now (caught by Python2.7 builds) is that downstream operations (such as pip install) on this file will fail if a subsequent run removes it to regenerate it (and the current session still wants to use it). We need to defend against this by probably making a temporary copy of this and feed that downstream. Another option would be to move it to that temp folder and create a symlink instead of copy (and when subsequent builds clean it follow up on symlinks and delete the target too); we would still need to feed back the temporary name of it. @asottile opinions?

Reply:
maybe I'm missing it, but couldn't the removal / creation just be moved into the critical section protected by the lock?

pre-commit has to do a similar thing: https://github.com/pre-commit/pre-commit/blob/c0b1f2ff25b53efcfa03098f2b4cb90ded905063/pre_commit/store.py#L79-L90

Reply:
should it make a dist for each environment then?

Reply:
For each session more likely. We need to keep around dist folder to not break peoples workflow in parallel.